### PR TITLE
Fix openssl deprecations

### PR DESCRIPTION
**Pull request by nolange**
This PR is trying to fix issues with openssl3 when compiled with `no-deprecated` or `no-engine`.
Conveniently its not needed to recompile openssl3, but it is enough to define the Macros `OPENSSL_NO_DEPRECATED` respectively  `OPENSSL_NO_ENGINE`.

Currently systemd compiles aslong the option `cryptolib=openssl` is not set. There is still work to do in `src/resolve/resolved-dns-dnssec.c`
AFAIU the code would need to use `rsa_pkey_from_n_e` `ecc_pkey_from_curve_x_y`, with possible some work on converting the parameters to a fitting `bignum`.

Fixes: #31958 and #31956

### CODE REVIEW

File: src/shared/openssl-util.c [hunk 1]
```diff
@@ -11,6 +11,15 @@
 #include "string-util.h"
 
 #if HAVE_OPENSSL
+#include <openssl/rsa.h>
+
+#if !defined(OPENSSL_NO_ENGINE) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#include <openssl/engine.h>
```
Comment (keszybz) [inline src/shared/openssl-util.c]: Maybe ident those those show that they are nested:
```C
#  include <openssl/rsa.h>
#    if !defined…
#      include <openssl/engine.h>
...
```

**Review by keszybz — COMMENTED** [review]
Looks reasonable.

File: src/resolve/resolved-dnstls-openssl.c [hunk 2]
```diff
@@ -392,8 +392,8 @@ int dnstls_manager_init(Manager *manager) {
 
         assert(manager);
 
-        ERR_load_crypto_strings();
-        SSL_load_error_strings();
+        OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
+        OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS | OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
```
Comment (keszybz) [inline src/resolve/resolved-dnstls-openssl.c]: Both functions return an error (1 on success or 0 on error), I think this should be checked.

Comment (keszybz) [reply src/resolve/resolved-dnstls-openssl.c]: And to answer the question whether we want this at all: sounds like we don't need to call this. So maybe just drop the whole thing altogether and see if tests pass?

File: src/resolve/resolved-dnstls-openssl.c [hunk 3]
```diff
@@ -392,8 +392,8 @@ int dnstls_manager_init(Manager *manager) {
```
Comment (yuwata) [inline src/resolve/resolved-dnstls-openssl.c:392]: Could you line break the commit message? It is too long per line.

File: src/shared/openssl-util.c (hunk 1 above)
Comment (nolange) [reply src/shared/openssl-util.c]: Fixed. Usually just try to fit in the surrounding code...

File: src/resolve/resolved-dnstls-openssl.c (hunk 3 above)
Comment (nolange) [reply src/resolve/resolved-dnstls-openssl.c:392]: Ok

**Comment by nolange** [comment]
I had some attempts trying to understand the dnssec stuff, not familiar enough to fix this and no unittest.

So this would be a fix for my issues, but not an allround cleanup for `NO_DEPRECATED`. Moving the includes out of the common header helped me finding the missing includes (which might or might not end up pulled in by other headers), probably gonna be a discussion whether this is necessary. My point would be that the more intricate OpenSSL internals shouldn't leak out of the implementations of the helper functions.

File: .github/workflows/build_test.sh [hunk 4]
```diff
@@ -130,17 +130,22 @@ ninja --version
...
     SECONDS=0
+    FCOMMON="-Werror"
 
     if [[ "$COMPILER" == clang && "$args" =~ Wmaybe-uninitialized ]]; then
         # -Wmaybe-uninitialized is not implemented in clang
         continue
     fi
 
+    if [[ "${CRYPTOLIB:?}" != openssl ]]; then
+        FCOMMON="$FCOMMON -DOPENSSL_NO_DEPRECATED"
```
Comment (bluca) [inline .github/workflows/build_test.sh]: It's fine to add a single job for this, but not to disable used APIs in all jobs, as we lose coverage of them. So change this to only affect a single job instead.

File: src/resolve/resolved-dns-dnssec.c [hunk 5]
```diff
@@ -14,12 +14,27 @@
...
-#if PREFER_OPENSSL && OPENSSL_VERSION_MAJOR >= 3
-#  pragma GCC diagnostic push
-#    pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#if PREFER_OPENSSL
+#  include <openssl/bn.h>
+#  include <openssl/rsa.h>
+#  include <openssl/ec.h>
+#  include <openssl/ecdsa.h>
+
+DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(EC_POINT*, EC_POINT_free, NULL);
```
Comment (bluca) [inline src/resolve/resolved-dns-dnssec.c]: This duplicates all these definitions many times over, which defeats the point of the common helper module, so please drop this commit

File: src/shared/openssl-util.h [hunk 6]
```diff
@@ -54,9 +53,7 @@ DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(SSL*, SSL_free, NULL);
 DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(BIO*, BIO_free, NULL);
 DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(EVP_MD_CTX*, EVP_MD_CTX_free, NULL);
 DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(ASN1_OCTET_STRING*, ASN1_OCTET_STRING_free, NULL);
-DISABLE_WARNING_DEPRECATED_DECLARATIONS;
-DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(ENGINE*, ENGINE_free, NULL);
```
Comment (bluca) [inline src/shared/openssl-util.h:58]: It was explicitly requested to have this definition in the header when it was added, so leave this here

Comment (nolange) [reply src/shared/openssl-util.h:58]: i think you mean a comment from @poettering ? aslong as no user is expected to need it I would leave it there.

File: src/resolve/resolved-dns-dnssec.c (hunk 5 above)
Comment (nolange) [reply src/resolve/resolved-dns-dnssec.c]: to me the helper module should provide the abstractions so those includes arent needed outside (sometime in the future). but i am not the maintainer, will do.

File: src/shared/openssl-util.h (hunk 6 above)
Comment (bluca) [reply src/shared/openssl-util.h:58]: Yes, it was asked to have it here regardless of whether it's used outside of this module, so please leave it here

File: .github/workflows/build_test.sh (hunk 4 above)
Comment (nolange) [reply .github/workflows/build_test.sh]: outside the "engines" no used apis are disabled. mostly stuff is changed between openssl 1.x and 3.x, so a test with 1.x would make most sense?

theoretical we could also dlsym the engine stuff, at which point nothing should change for unit tests.

File: src/shared/openssl-util.h (hunk 6 above)
Comment (nolange) [reply src/shared/openssl-util.h:58]: well, lets ping @poettering, if this specific instance is really an issue?
it seemed more like a question to me.

File: .github/workflows/build_test.sh (hunk 4 above)
Comment (bluca) [reply .github/workflows/build_test.sh]: The engines interface is exactly what we need coverage for

File: src/shared/openssl-util.h (hunk 6 above)
Comment (nolange) [reply src/shared/openssl-util.h:58]: Link to comment is here: https://github.com/systemd/systemd/pull/31261#discussion_r1484311058

The definition is only (to be) used via `openssl_load_key_from_token`, is conditional and deprecated,should not end up in the header-

File: .github/workflows/build_test.sh [hunk 7]
```diff
@@ -130,17 +130,22 @@ ninja --version
 
 for args in "${ARGS[@]}"; do
     SECONDS=0
+    FCOMMON="-Werror"
```
Comment (keszybz) [inline .github/workflows/build_test.sh]: This is not right. It'll make things fail whenever any warning is emitted, making CI very flakey.

File: src/shared/openssl-util.h (hunk 6 above)
Comment (yuwata) [reply src/shared/openssl-util.h:58]: Even it is requested in the linked comment, but that's not necessary in the header, and keeping the functions in header requires further ifdef and include in the header. That's annoying. I agree with @nolange.

File: .github/workflows/build_test.sh (hunk 7 above)
Comment (yuwata) [reply .github/workflows/build_test.sh]: I cannot follow. The flag has been already set unconditionally.

File: .github/workflows/build_test.sh (hunk 4 above)
Comment (yuwata) [reply .github/workflows/build_test.sh]: Yes. @bluca is right. Please introduce a new test job with the flag.

File: src/shared/openssl-util.h (hunk 6 above)
Comment (poettering) [reply src/shared/openssl-util.h:58]: yeah, with the ifdeffery might make more sense to move this specific case if it's really not needed anywhere else, and is unlikely to be.

**Comment by poettering** [comment]
lgtm. 

@mrc0mmand any chance you can comment on the CI changes this makes?

File: .github/workflows/build_test.sh (hunk 7 above)
Comment (mrc0mmand) [reply .github/workflows/build_test.sh]: Yeah, we already have `--werror` in the meson call below, so this should be dropped.

**Comment by poettering** [comment]
@nolange can you rework this to drop the werror thing?

File: .github/workflows/build_test.sh (hunk 4 above)
Comment (bluca) [reply .github/workflows/build_test.sh]: This is still unresolved. Engines with Openssl3 must be covered by CIs, as it's the only realistic way to use pkcs11 and other hardware stuff with our tools.

**Review by bluca — APPROVED** [review]

**Pull request merged**